eth/executionclient/multi_client.go

```diff
@@ -0,0 +1,404 @@
+	errCh := make(chan error, 1)
+
+	var singleLogsCh <-chan BlockLogs
+	var singleErrCh <-chan error
+	var nothingToSyncCount atomic.Int64
+
+	// Find a client that's able to provide the requested data.
+	startFetchingFunc := func(client SingleClientProvider) (any, error) {
+		cl, ce, err := client.FetchHistoricalLogs(ctx, fromBlock)
+		if err != nil {
+			if errors.Is(err, ErrNothingToSync) {
+				nothingToSyncCount.Add(1)
+			}
+			// Consider ErrNothingToSync as an error to make sure that other nodes return ErrNothingToSync too.
+			// If they don't, it means that the current node is out of sync.
+			// Therefore, we keep count of how many ErrNothingToSync we saw.
+			return nil, err
+		}
+
+		singleLogsCh = cl
+		singleErrCh = ce
+		return nil, nil
+	}
+
+	if _, err := ec.call(ec.setMethod(ctx, "FetchHistoricalLogs [start fetching]"), startFetchingFunc); err != nil {
+		if int(nothingToSyncCount.Load()) == len(ec.clients) {
+			// All clients returned ErrNothingToSync
+			return nil, nil, ErrNothingToSync
+		}
+		return nil, nil, err
+	}
+
+	go func() {
+		defer func() {
+			close(logsCh)
+			close(errCh)
+		}()
+
+		var needInitChans atomic.Bool
+
+		// Getting historical logs may take significant amount of time.
+		// The current client may fail in the middle of the process,
+		// and also some clients may become available again, therefore we need to try other clients on error.
+		// Since singleLogsCh and singleErrCh are initialized by the active client,
+		// we need to re-initialize them if we try other clients.
+		processLogsFunc := func(client SingleClientProvider) (any, error) {
+			if needInitChans.Load() {
+				if _, err := startFetchingFunc(client); err != nil {
+					return nil, err
+				}
+			} else {
+				needInitChans.Store(true)
+			}
+
+			var lastBlock uint64
+			for {
+				select {
+				case <-ctx.Done():
+					return nil, ctx.Err()
+				case <-ec.closed:
+					return nil, fmt.Errorf("client closed")
+				case log, ok := <-singleLogsCh:
+					if !ok {
+						// Underlying channel is closed -> no more logs.
+						return nil, nil
+					}
+					logsCh <- log
+					lastBlock = max(lastBlock, log.BlockNumber)
+				case err, ok := <-singleErrCh:
+					if !ok {
+						// If the error channel closed, treat that as no more logs or success.
+						return nil, nil
+					}
+					fromBlock = max(fromBlock, lastBlock+1)
+					return nil, err // Try another client
+				}
+			}
+		}
+
+		if _, err := ec.call(ec.setMethod(ctx, "FetchHistoricalLogs [process logs]"), processLogsFunc); err != nil {
+			errCh <- err
+		}
+	}()
+
+	return logsCh, errCh, nil
+}
+
+// StreamLogs subscribes to events emitted by the contract.
+func (ec *MultiClient) StreamLogs(ctx context.Context, fromBlock uint64) <-chan BlockLogs {
+	logs := make(chan BlockLogs)
+
+	go func() {
+		defer close(logs)
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-ec.closed:
+				return
+			default:
+				f := func(client SingleClientProvider) (any, error) {
```

Comment:
this can be defined outside of the loop. even though compiler/runtime takes care of it anyway.

Reply:
It needs to be a closure because `fromBlock` is updated inside

Reply:
I added a comment